### DISCUSSION TITLE
Multiple devs on single linux server via Remote - SSH?

### DISCUSSION THREAD

**Discussion by aschweig** [scalameta/metals#4881] [Q&A]
I would like to use vscode Metals along with the Remote - SSH extension.  We have a powerful development linux box that several developers share.

With defaults, we have only been successful with a single developer at-a-time running vscode Remote - SSH with remote metals.
We suspect the root cause is the following [restriction from the bloop](https://scalacenter.github.io/bloop/docs/what-is-bloop#limitations) server:

> There can only be one bloop server instance running per machine

The default port for bloop is 8212 -- does the [above limitation goes deeper than the port number](https://github.com/scalameta/metals/pull/2267)?

Again, using defaults, when I try to have a second user start metals via Remote - SSH, the logs show something like this:

```
2023.01.20 09:23:21 INFO  tracing is disabled for protocol LSP, to enable tracing of incoming and outgoing JSON messages create an empty file at /mnt/homes/aschweig/Project$
2023.01.20 09:23:21 INFO  logging to file /mnt/homes/aschweig/Project/Foo/.metals/metals.log
2023.01.20 09:23:21 INFO  Started: Metals version 0.11.10 in workspace '/mnt/homes/aschweig/Project/Foo' for client Visual Studio Code 1.74.2.
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions 1.7.x or earlier.
SLF4J: Ignoring binding found at [jar:file:/home/aschweig/.cache/coursier/v1/https/repo1.maven.org/maven2/com/outr/scribe-slf4j_2.13/3.10.5/scribe-slf4j_2.13-3.10.5.jar!/org/s$
SLF4J: See https://www.slf4j.org/codes.html#ignoredBindings for an explanation.
2023.01.20 09:23:23 INFO  time: initialize in 2.31s
2023.01.20 09:23:23 INFO  Attempting to connect to the build server...
2023.01.20 09:23:23 INFO  Bloop uses /mnt/homes/aschweig/jdk8u282-b08/jre defined at /home/aschweig/.bloop/bloop.json
2023.01.20 09:23:24 INFO  skipping build import with status 'Installed'
2023.01.20 09:23:24 ERROR Starting the bsp launcher for bloop...
2023.01.20 09:23:24 ERROR Opening a bsp server connection with 'bsp --protocol local --socket /tmp/bsp-launcher1245002761402780631/bsp.socket'...
2023.01.20 09:23:24 ERROR Waiting for the bsp connection to come up...
2023.01.20 09:23:24 ERROR error: The command bsp --protocol local --socket /tmp/bsp-launcher1245002761402780631/bsp.socket returned with an error
2023.01.20 09:23:24 ERROR > [W] Scala instance jar /home/aschweig/.cache/coursier/v1/https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-emoji/0.42.12/flexmark-ext-emoji-0.42.12.jar doesn't exist!
...
```

One approach I found is to set unique values for metals.bloop-port for each user, e.g., in their ``.bash_profile`` on the dev server:
```
export JAVA_TOOL_OPTIONS="-Dmetals.bloop-port=8213"
```

Is there a better or standard way to have multiple users use Remote - SSH on a single linux dev box?  Can all users share a single bloop instance?  If so, how?

**Comment by tgodzik** [comment]
Thanks for the question and sorry for not responding earlier. This is something that we need to improve on the Bloop side, which is already implemented in one of the forks and we need to merge it back. Namely, it will used named sockets instead of a well known port. Your workaround is probably the only workaround currently. Alternatively, if you are using mill or sbt you can try using them as BSP servers, which will be started once per workspace.
